userdocs/diagnostics/branches-have-mismatching-types.md

```diff
@@ -0,0 +1,35 @@
+# Branches have mismatching types
+
+When an `if` or `switch` statement is used as an expression, each of its branches must produce a value of the same type when type checked independently.
+
+For example, the following code is invalid because one branch of the `if` expression produces a value of type `String`, and the other branch produces a value of type `Int`. As a result, the compiler is unable to infer a type for `result`.
```

Comment:
"the compiler is unable to infer a type for `result`" sounds like a compiler limitation rather than a fundamental issue that the programmer has to fix. Maybe something like

> The type of `result` can be either `Int` or `String`, but not both.

We should also communicate how to resolve the error. In this case, either you need to adjust the branches to produce a matching type, or if you want to produce different types based on a dynamic condition, then you should type erase to an `any` type by writing a type annotation.